### PR TITLE
Move map management to `BpfBytecode`

### PR DESCRIPTION
**Pull request by viktormalik**
The current map management using `MapManager`/`IMap`/`Map`/`FakeMap` is quite difficult to read in the code as map settings are scattered across multiple files. This PR attempts to unify everything in `BpfBytecode`, where it logically belongs.

Maps are stored by their name (map ids are not used anymore) and represented using a new `BpfMap` class. The class contains `struct bpf_map` which is libbpf map representation containing basic map information (map type, max entries, key and value size) which libbpf is able to parse from BTF stored in the generated ELF. In addition, `BpfMap` stores map fd. Once we let libbpf auto-create maps for us in future, we will be able to drop manual map creation and fd will also be stored in `struct bpf_map`. Logically, the new `BpfMap` class replaces the previous `Imap`/`Map` classes.

The PR migrates all map usages to the new map management. The migration had to deal with two kinds of maps:
- Internal maps which usually only use fd and therefore the migration was mostly straightforward.
- Async map actions (`print`, `clear`, and `zero`) which came with two problems: (1) async actions need map ids which were dropped and (2) some information needed (e.g. key/value type) is not stored in `BpfMap`.  Since I wanted to keep the generated BPF programs decoupled from runtime as much as possible, I decided to store the above information in `RequiredResources` which are accessible by async actions. Part of the information was already there and the PR adds map name<->id mappings (for maps used inside async actions).

Finally, `MapManager`, `IMap`, `Map`, and `FakeMap` are dropped.

See individual commit messages for more details.

Note: this is the last part of #2911. It's quite big but unfortunately, I wasn't able to split it into smaller parts as it wouldn't make much sense to me. If you still prefer to post it per-commit, let me know.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by viktormalik** [comment]
For some reason, there are random failures of the `call.ustack_stack_mode_env*` runtime tests. Restarting the job usually helps but I don't remember seeing such failures before so they may be caused by something in this PR. I'll try to have a look.

**Comment by viktormalik** [comment]
> For some reason, there are random failures of the `call.ustack_stack_mode_env*` runtime tests. Restarting the job usually helps but I don't remember seeing such failures before so they may be caused by something in this PR. I'll try to have a look.

Unfortunately, I'm not able to reproduce the issue locally (even when using Nix).

**Comment by ajor** [comment]
What kinds of failures were they?

**Review by ajor — COMMENTED** [review]
Overall looks good to me.

There do seem to be a lot of map lookups going on throughout the whole process, which might be slowing us down more than necessary.

e.g. in the output path (where performance counts), we have a multi-step process to obtain a BpfMap object:
1. Ringbuffer -> `print->mapid`
2. `resources.maps_by_id[mapid]` -> `map_name`
3. `bytecode.getMap(map_name)` -> `BpfMap`

Also `resources.map_key` and `resources.map_type` being two separate map lookups.

**Comment by viktormalik** [comment]
> What kinds of failures were they?

IIRC there was no stack printed (just empty output).

File: src/bpfmap.cpp
Comment (viktormalik) [reply src/bpfmap.cpp:20]: Agreed, let's use standard `uint32_t`.

File: src/bpftrace.cpp
Comment (viktormalik) [reply src/bpftrace.cpp:418]: Yes, it's an internal error which should never happen.

File: src/required_resources.cpp
Comment (viktormalik) [reply src/required_resources.cpp:20]: Good point. I guess that we could remove it from `libruntime` and just move it to AOT. I didn't do it b/c `required_resources.h` was included by many files in `libruntime` so I wanted to have them coupled together. But it probably doesn't make any sense.

**Comment by viktormalik** [comment]
> Overall looks good to me.
> 
> There do seem to be a lot of map lookups going on throughout the whole process, which might be slowing us down more than necessary.

That's a good point, although I'm wondering if we can really avoid that.

> e.g. in the output path (where performance counts), we have a multi-step process to obtain a BpfMap object:
> 
> 1. Ringbuffer -> `print->mapid`

This is necessary b/c mapid is stored in codegen. I don't see a way to avoid this.

> 2. `resources.maps_by_id[mapid]` -> `map_name`
> 3. `bytecode.getMap(map_name)` -> `BpfMap`

Probably the most efficient would be to have mapid -> `BpfMap` mapping but the problem is that we need mapid *before* codegen but only create `BpfMap` *after* codegen. We could create an extra mapid -> `BpfMap` mapping (perhaps in `BpfBytecode`) after `BpfMap` objects are created. I can give it a try. If you have other idea how to resolve this, let me know.

> Also `resources.map_key` and `resources.map_type` being two separate map lookups.

That's a good point. We could probably aggregate them in some structure like `MapInfo` in `RequiredResources` (together with other info such as map ids, (l)hist args, etc.).

**Comment by viktormalik** [comment]
Thanks for the review @ajor! I implemented all of the comments, noticeable changes in the new version are:
- moved linking of `required_resources.cpp` out of `libruntime` to `libaot` only,
- aggregated map information in `RequiredResources` to a single struct,
- created a new map `BpfBytecode::maps_by_id_` to allow efficient lookup of `BpfMap` objects by map ids.

File: src/bpfbytecode.cpp
Comment (viktormalik) [inline src/bpfbytecode.cpp:131]: IIUC, this should be fine b/c references to `std::map` are not invalidated by inserting new elements. But @ajor is the C++ expert here so I'd like to highlight this so that you double check.

Comment (ajor) [reply src/bpfbytecode.cpp:131]: Yep, looks fine to me 👍

File: src/bpftrace.cpp
Comment (ajor) [inline src/bpftrace.cpp:1575]:
```suggestion
    const auto &map_key = map_info.key;
```

File: src/bpftrace.cpp
Comment (ajor) [inline src/bpftrace.cpp:1674]:
```suggestion
  const auto &map_key = resources.maps_info.at(map.name()).key;
```

File: src/output.cpp
Comment (ajor) [inline src/output.cpp:415]:
```suggestion
  const auto &map_type = map_info.value_type;
```

**Review by ajor — APPROVED** [review]
Looks great!

Just spotted few extra copies we can eliminate. I'm extra suspicious whenever I see `auto x = ` because it's easy to hide these expensive copies (not saying that I don't code like this myself!)

**Comment by viktormalik** [comment]
> Just spotted few extra copies we can eliminate. I'm extra suspicious whenever I see `auto x = ` because it's easy to hide these expensive copies (not saying that I don't code like this myself!)

Thanks, those are great catches! I try to avoid `auto x = `, too, but sometimes my brain falls into a just-auto-everything mode and that's when these happen :laughing:

Anyways, I'd like to wait for @danobi's review before merging this.

**Review by danobi — COMMENTED** [review]
only looked at first 3 commits. might get more time this weekend to finish

File: src/bpfbytecode.h
Comment (viktormalik) [reply src/bpfbytecode.h:63]: This is on purpose to get the maps in alphabetical order when printing them at the end.

File: src/bpfbytecode.cpp
Comment (viktormalik) [reply src/bpfbytecode.cpp]: Good idea, I didn't know about this feature. Done

File: src/bpfbytecode.h
Comment (danobi) [inline src/bpfbytecode.h:40]: `interal_type` -> `internal_type`

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [inline src/ast/passes/codegen_llvm.cpp:3338]: nit: could destructure here. eg:

```
for (const auto &[name, info] : resources.maps_info) {
    ...
}
```

File: src/required_resources.cpp
Comment (danobi) [inline src/required_resources.cpp]: Unrelated, but wondering if we should always use `LOG(FATAL)` instead of exceptions, since they serve similar purposes as documented in our coding guidelines. Would be an easy codemod

File: src/required_resources.h
Comment (danobi) [inline src/required_resources.h:56]: Need to add to `archive()` ?

File: src/bpfbytecode.h
Comment (danobi) [inline src/bpfbytecode.h:64]: If the map is non-nullable here, maybe better to store a reference?

File: src/CMakeLists.txt
Comment (danobi) [inline src/CMakeLists.txt:38]: This means from core bpftrace files we only use required_resources.h definitions. Meaning if future pubic methods are exposed and called from core bpftrace, then you'd get a linker error then.

What was the linker error that motivated this change? Curious cuz I feel like it should work.

File: src/aot/CMakeLists.txt
Comment (danobi) [inline src/aot/CMakeLists.txt:1]: Yeah this is a bit unfortunate. I feel like cmakelists.txt should try not to reach up to parent  or sibling dirs

**Review by danobi — COMMENTED** [review]
few comments, but great cleanup!

File: src/required_resources.h
Comment (viktormalik) [reply src/required_resources.h:56]: Yes, good catch.

File: src/ast/passes/codegen_llvm.cpp
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp:3338]: Love this syntax! Would be nice to start using it more across the codebase.

File: src/bpfbytecode.h
Comment (viktormalik) [reply src/bpfbytecode.h:64]: I don't think it's possible to store a reference in a `std::map`. We could use `std::reference_wrapper` but that feels like adding an unnecessary complexity to me.

File: src/CMakeLists.txt
Comment (viktormalik) [reply src/CMakeLists.txt:38]: Since no function from `required_resources.cpp` is used by non-aot code, the linker optimizes it away when linking `libruntime` and afterwards linking `libaot` fails with:
```
/usr/bin/ld: aot/libaot.a(aot.cpp.o): in function `bpftrace::aot::(anonymous namespace)::load_required_resources(bpftrace::BPFtrace&, unsigned char*, unsigned long)':
/home/vmalik/src/bpf/bpftrace/src/aot/aot.cpp:77: undefined reference to `bpftrace::RequiredResources::load_state(unsigned char const*, unsigned long)'
/usr/bin/ld: aot/libaot.a(aot.cpp.o): in function `bpftrace::aot::(anonymous namespace)::generate_btaot_section(bpftrace::RequiredResources const&, bpftrace::BpfBytecode const&)':
/home/vmalik/src/bpf/bpftrace/src/aot/aot.cpp:109: undefined reference to `bpftrace::RequiredResources::save_state(std::ostream&) const'
```
My original solution was to add a dummy constructor of `RequiredResrouces` to `required_resources.cpp` but @ajor had a good point that it doesn't make sense linking an object file from which no code is used. OTOH, if we ever added some code needed from `libruntime`, we'd have to revert all this.

File: src/aot/CMakeLists.txt
Comment (viktormalik) [reply src/aot/CMakeLists.txt:1]: I don't like it either. But I couldn't come with any other solution, besides adding the dummy constructor (see the other thread).

File: src/required_resources.cpp
Comment (viktormalik) [reply src/required_resources.cpp]: This is something I'm often confused about. In some places, we use `LOG(FATAL)`, in others we raise `std::runtime_exception`. It'd be nice to unify this, perhaps we can discuss on the next office hours.

File: src/aot/CMakeLists.txt
Comment (ajor) [reply src/aot/CMakeLists.txt:1]: Should required_resources.cpp live in the aot directory, since that's where it's used?

Comment (viktormalik) [reply src/aot/CMakeLists.txt:1]: Possibly but we wouldn't have `required_resources.cpp` next to `required_resources.h`.

Comment (ajor) [reply src/aot/CMakeLists.txt:1]: I was thinking of moving the header too, but I suppose it's used in more than just AOT so that doesn't make sense.

How about creating a new CMake library target just for required_resources? Then both runtime and aot can link against that. Would it solve the unused symbol issues?

File: src/bpfbytecode.h
Comment (danobi) [reply src/bpfbytecode.h:64]: Ah true, yeah. Ptr is fine

File: src/required_resources.cpp
Comment (danobi) [reply src/required_resources.cpp]: Sounds good!

File: src/aot/CMakeLists.txt
Comment (danobi) [reply src/aot/CMakeLists.txt:1]: Agreed - separate target is probably cleanest solution

Comment (viktormalik) [reply src/aot/CMakeLists.txt]: Good idea, done.

File: src/ast/passes/codegen_llvm.cpp
Comment (danobi) [reply src/ast/passes/codegen_llvm.cpp]: I suspect clang-tidy can probably catch this. I'll make a note in the tracking issue and take a look when I get a chance

**Review by danobi — APPROVED** [review]

**Comment by viktormalik** [comment]
Rebased. Going to merge this to avoid more conflicts with open PRs.

**Pull request merged**
